rfd/0212-jsonpath-interpolation.md

```diff
@@ -0,0 +1,584 @@
+  name: distributed-idp
+spec:
+  priority: 0
+  traits_map:
+    logins:
+      # evaluates to ["alice"]
+      - jsonpath(external.groups, "$.logins")
+    env:
+      # evaluates to ["staging", "dev"]
+      - jsonpath(external.groups, "$.env")
+```
+
+These traits can then be used in role templates, label expressions, etc.
+
+```yaml
+kind: role
+version: v7
+metadata:
+  name: template
+spec:
+  allow:
+    logins: '{{external.logins}}'
+    node_labels_expression: 'contains(external.env, labels["env"])'
+```
+
+### UX
+
+#### User stories
+
+The user stories below will explore example custom OIDC solutions with
+potential Teleport configurations to consume the custom OIDC claims using the
+new `jsonpath` function.
+
+##### Example: IdP with arbitrary json claims
+
+Let's say we have a custom IdP which directly supports arbitrary json claims to
+be set for users. Below is an example claim object for user `alice`.
+
+```json
+{
+  "groups": {
+    "teleport": {
+      "roles": ["template"],
+      "node": {
+        "logins": "alice",
+        "labels": {
+          "*": "*"
+        }
+      },
+      "app": {
+        "labels": {
+          "env": "staging"
+        }
+      }
+    }
+  }
+}
+```
+
+The administrator wants to take these claims and map them directly to role
+conditions using [role templating](https://goteleport.com/docs/admin-guides/access-controls/guides/role-templates/).
+
+First, the admin needs to create a `login_rule` to map this arbitrary json
+object into a set of user traits.
+
+```yaml
+kind: login_rule
+version: v1
+metadata:
+  name: arbitrary-json-idp
+spec:
+  priority: 0
+  traits_map:
+    roles:
+      # evaluates to ["template"]
+      - jsonpath(external.groups, "$.teleport.roles")
+    logins:
+      # evaluates to ["alice"]
+      - jsonpath(external.groups, "$.teleport.node.logins")
+    node_labels_*:
+      # evaluates to "*"
+      - jsonpath(external.groups, "$.groups.teleport.node.labels['*']")
+    node_labels_env:
+      # evaluates to []
+      - jsonpath(external.groups, "$.groups.teleport.node.labels.env")
+    app_labels_*:
+      # evaluates to []
+      - jsonpath(external.groups, "$.groups.teleport.app.labels['*']")
+    app_labels_env:
+      # evaluates to "staging"
+      - jsonpath(external.groups, "$.groups.teleport.app.labels.env")
+```
+
+The mapped traits can now be used as if they are standard OIDC claims in the
+OIDC connector's `claims_to_roles` spec to assign the template role to the user.
+
+```yaml
+kind: oidc 
+version: v2
+metadata:
+  name: arbitrary-json-idp
+spec:
+  ...
+  claims_to_roles:
+    - claim: "roles"
+      value: "template"
+      roles: ["template"]
+```
+
+Lastly, the admin can create the template role and utilize the mapped traits.
+
+```yaml
+kind: role
+version: v7
+metadata:
+  name: template
+spec:
+  allow:
+    logins: '{{external.logins}}'
+    node_labels:
+      '*': '{{external.node_labels_*}}'
+      'env': '{{external.node_labels_env}}'
```

Comment:
wouldn't it be nice if we didn't have to statically list all the label keys... although I guess one might be able to use a node_labels_expression for that

Reply:
oh i guess this is addressed with the jsonpath-plus syntax